src/targets/gpu/fuse_mlir.cpp

```diff
@@ -878,6 +878,50 @@ struct find_pointwise_mlir
     }
 };
 
+struct find_unpack_int4_mlir_op
+{
+    auto matcher() const
+    {
+        return match::name("gpu::mlir_op")(
+            match::any_of[match::inputs()](match::name("unpack_int4").bind("unpack_int4")));
+    }
+
+    void apply(module_pass_manager& mpm, const match::matcher_result& r) const
+    {
+        auto ins      = r.result;
+        auto* mm      = ins->module_inputs().front();
+        module_ref nm = mpm.create_module("int4:" + mm->name());
+        nm->set_bypass();
+
+        std::vector<instruction_ref> x_in;
+        std::unordered_map<instruction_ref, instruction_ref> map_ins;
+        int ct = 0;
+
+        for(auto inp : ins->inputs())
+        {
+            instruction_ref t_ins;
+            if(inp->get_operator().name() == "unpack_int4")
+            {
+                auto chld = inp->inputs()[0];
+                t_ins     = nm->add_parameter(chld->name() + std::to_string(++ct),
```

Comment:
Dont use `to_string` to make the parameters names as that wont work if there is more than 10 parameters. Instead use `param_name(ct++)`.